libphal/phal_dump.C

```diff
@@ -0,0 +1,418 @@
+} __attribute__((packed));
+
+/**
+ * @brief Return the proc target corresponding to failing unit
+ * @param[in] failingUnit Position of the proc containing the failed SBE
+ *
+ * @return Pointer to the pdbg target for proc
+ *
+ * Exceptions: PDBG_TARGET_NOT_OPERATIONAL if target is not available
+ */
+struct pdbg_target* getProcFromFailingId(const uint32_t failingUnit)
+{
+	struct pdbg_target* proc = NULL;
+	struct pdbg_target* target;
+	pdbg_for_each_class_target("proc", target)
+	{
+		if (pdbg_target_index(target) != failingUnit) {
+			continue;
+		}
+
+		if (pdbg_target_probe(target) != PDBG_TARGET_ENABLED) {
+			break;
+		}
+		proc = target;
+	}
+	if (proc == NULL) {
+		log(level::ERROR,
+		    "No Proc target found to execute the dump for failing "
+		    "unit=%d",
+		    failingUnit);
+		throw sbeError_t(exception::PDBG_TARGET_NOT_OPERATIONAL);
+	}
+	return proc;
+}
+
+/**
+ * @brief Initialize the PDBG and SBE to start collection
+ * @param[in] failingUnit Position of the proc containing the failed SBE
+ *
+ * Exceptions: PDBG_INIT_FAIL for any pdbg init related failure.
+ *             PDBG_TARGET_NOT_OPERATIONAL for invalid failing unit
+ *             HWP_EXECUTION_FAILED if the extract rc procedure is failing
+ */
+struct pdbg_target* preCollection(const uint32_t failingUnit)
+{
+	// Initialize PDBG with KERNEL backend
+	pdbg::init(PDBG_BACKEND_KERNEL);
+
+	// FSI and PIB targets for executing HWP
+	struct pdbg_target *fsi, *pib;
+	char path[16];
+
+	// Find the proc target from the failing unit id
+	struct pdbg_target* proc = getProcFromFailingId(failingUnit);
+
+	// Probe FSI for HWP execution
+	sprintf(path, "/proc%d/fsi", pdbg_target_index(proc));
+	fsi = pdbg_target_from_path(NULL, path);
+	if (fsi == nullptr) {
+		log(level::ERROR, "Failed to get FSI target for(%s)",
+		    pdbg_target_path(proc));
+		throw dumpError_t(exception::PDBG_TARGET_NOT_OPERATIONAL);
+	}
+
+	// Probe PIB for HWP execution
+	sprintf(path, "/proc%d/pib", pdbg_target_index(proc));
+	pib = pdbg_target_from_path(NULL, path);
+	if (pib == nullptr) {
+		log(level::ERROR, "Failed to get PIB target for(%s)",
+		    pdbg_target_path(proc));
+		throw dumpError_t(exception::PDBG_TARGET_NOT_OPERATIONAL);
+	}
+
+	if ((pdbg_target_probe(fsi) != PDBG_TARGET_ENABLED ||
+	     pdbg_target_probe(pib) != PDBG_TARGET_ENABLED)) {
+		log(level::ERROR, "Failed to prob PIB or FSI");
+		throw dumpError_t(exception::PDBG_TARGET_NOT_OPERATIONAL);
+	}
+
+	// Execute SBE extract rc to set up sdb bit for pibmem dump to work
+	// TODO Add error handling or revisit procedure later
+	fapi2::ReturnCode fapiRc;
+	P10_EXTRACT_SBE_RC::RETURN_ACTION recovAction;
+	fapiRc = p10_extract_sbe_rc(proc, recovAction, true);
+	if (fapiRc != fapi2::FAPI2_RC_SUCCESS) {
+		log(level::ERROR,
+		    "Failed in p10_extract_sbe_rc for proc=%s, rc=0x%08X",
+		    pdbg_target_path(proc), fapiRc);
+		throw dumpError_t(exception::HWP_EXECUTION_FAILED);
+	}
+
+	log(level::INFO, "SBE RC=0x%08X", recovAction);
```

Comment:
Log function deosn't keep function tag,"SBE RC" change to SBE Recovery action: " ?